### PR TITLE
Fix building clad for conda builds

### PR DESCRIPTION
**Pull request by chrisburr**

### CODE REVIEW

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by vgvassilev** [comment]
We get:
```
10:31:27 CMake Error at interpreter/cling/tools/plugins/clad/CMakeLists.txt:15 (list):
10:31:27   list does not recognize sub-command JOIN
```

**Comment by chrisburr** [comment]
The builds are failing as the version of cmake doesn't support `list(JOIN`. Do you have a preferred workaround or alternate way of fixing this?

File: interpreter/CMakeLists.txt [hunk 1]
```diff
@@ -390,6 +382,13 @@ if (builtin_clang)
     CACHE STRING "Clang include directories.")
 endif()
 
+if (builtin_clang)
+  add_subdirectory(llvm/src/tools/clang EXCLUDE_FROM_ALL)
+else()
+  add_subdirectory(cling)
+endif()
+endif(builtin_llvm)
+
```
Comment (vgvassilev) [inline interpreter/CMakeLists.txt]: I am wondering what's the observable change behavior of this change.

File: interpreter/cling/tools/plugins/clad/CMakeLists.txt [hunk 2]
```diff
@@ -12,6 +12,7 @@ set(CLADDIFFERENTIATOR_LIB
   ${_clad_byproduct_binary_dir}/lib/Differentiator/${CMAKE_STATIC_LIBRARY_PREFIX}cladDifferentiator${CMAKE_STATIC_LIBRARY_SUFFIX}
   ${_clad_byproduct_binary_dir}/tools/${CMAKE_STATIC_LIBRARY_PREFIX}cladPlugin${CMAKE_STATIC_LIBRARY_SUFFIX}
   )
+list(JOIN CLANG_INCLUDE_DIRS "|" _clad_clang_include_dirs)
```
Comment (vgvassilev) [inline interpreter/cling/tools/plugins/clad/CMakeLists.txt]: I am not sure what we are solving here. What is `_clad_clang_include_dirs`? If we are allowing CLANG_INCLUDE_DIRS to be different from `_clad_clang_include_dirs` that'd be incorrect.

Comment (chrisburr) [reply interpreter/cling/tools/plugins/clad/CMakeLists.txt]: Disclaimer: I'm not very knowledgeable about cmake so I might have misunderstood how this patch actually works.

The issue is that `DCLANG_INCLUDE_DIRS` can contain multiple items such as `/my/dir1` `/my/dir2` so the `CMAKE_ARGS` argument is expanded to:

```cmake
   CMAKE_ARGS -G ${CMAKE_GENERATOR} -DCLAD_BUILD_STATIC_ONLY=ON
               -DCMAKE_INSTALL_PREFIX=${CLING_PLUGIN_INSTALL_PREFIX}
               -DCLAD_PATH_TO_LLVM_BUILD=${LLVM_BINARY_DIR}
               -DCLANG_INCLUDE_DIRS=/my/dir1 /my/dir2
```

And `/my/dir2` isn't passed through as an include directory when configuring `clad`.

This solution works by setting `_clad_clang_include_dirs` to `/my/dir1|/my/dir2` and then using `LIST_SEPARATOR |` to tell `ExternalProject_Add` to split up the lists.

File: interpreter/CMakeLists.txt (hunk 1 above)
Comment (chrisburr) [reply interpreter/CMakeLists.txt]: I think `CLANG_INCLUDE_DIRS` needs to be set before doing `add_subdirectory(cling)`.

File: interpreter/cling/tools/plugins/clad/CMakeLists.txt (hunk 2 above)
Comment (vgvassilev) [reply interpreter/cling/tools/plugins/clad/CMakeLists.txt]: Don't you need to just wrap it into quotes? Eg. `-DCLANG_INCLUDE_DIRS="${CLANG_INCLUDE_DIRS}`

Comment (vgvassilev) [reply interpreter/cling/tools/plugins/clad/CMakeLists.txt]: Ok, could you use `string(REPLACE...`, instead of JOIN?

Comment (amadio) [reply interpreter/cling/tools/plugins/clad/CMakeLists.txt]: `list(JOIN ...)` is only available with CMake 3.12 and above, we cannot use it, as we need to support CMake 3.4.3 and above in ROOT.

Comment (chrisburr) [reply interpreter/cling/tools/plugins/clad/CMakeLists.txt]: Changed 👍

File: interpreter/CMakeLists.txt (hunk 1 above)
Comment (vgvassilev) [reply interpreter/CMakeLists.txt]: I am a little confused, how did it work before?

File: interpreter/cling/tools/plugins/clad/CMakeLists.txt [hunk 3]
```diff
@@ -12,6 +12,7 @@ set(CLADDIFFERENTIATOR_LIB
   ${_clad_byproduct_binary_dir}/lib/Differentiator/${CMAKE_STATIC_LIBRARY_PREFIX}cladDifferentiator${CMAKE_STATIC_LIBRARY_SUFFIX}
   ${_clad_byproduct_binary_dir}/tools/${CMAKE_STATIC_LIBRARY_PREFIX}cladPlugin${CMAKE_STATIC_LIBRARY_SUFFIX}
   )
+string(REPLACE ";" "|" _clad_clang_include_dirs "${CLANG_INCLUDE_DIRS}")
```
Comment (vgvassilev) [inline interpreter/cling/tools/plugins/clad/CMakeLists.txt]: Could you add a comment why we need that?

File: interpreter/CMakeLists.txt (hunk 1 above)
Comment (chrisburr) [reply interpreter/CMakeLists.txt]: I have no idea, but I just double checked removing this change and [building clad fails]("https://gitlab.cern.ch/cburr/root-conda/-/jobs/2832171 ) (search for `Command failed: 2`).

**Comment by Axel-Naumann** [comment]
@phsft-bot build!

**Comment by Axel-Naumann** [comment]
These errors seem genuine? @amadio might this be? (I don't see these errors in the current master.)

**Comment by chrisburr** [comment]
There might be an issue with the way I'm using `string(REPLACE` in place of `list(JOIN` in
260b2cc, I'll check.

**Pull request closed without merging**